File: src/main/java/net/slqmy/better_projectiles_plugin/listener/NuclearGhastDeathListener.java
package net.slqmy.better_projectiles_plugin.listener;

import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Ghast;
import org.bukkit.entity.ThrownPotion;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDeathEvent;
import org.bukkit.persistence.PersistentDataContainer;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;
import org.jetbrains.annotations.NotNull;

import net.slqmy.better_projectiles_plugin.BetterProjectilesPlugin;

public class NuclearGhastDeathListener implements Listener {

    private final BetterProjectilesPlugin plugin;

    public NuclearGhastDeathListener(BetterProjectilesPlugin plugin) {
        this.plugin = plugin;
    }

    @EventHandler
    public void onNuclearGhastDeath(@NotNull EntityDeathEvent event) {
        if (event.getEntity() instanceof Ghast ghast) {
            PersistentDataContainer container = ghast.getPersistentDataContainer();

            boolean isNuclearGhast = Boolean.TRUE.equals(container.get(plugin.getNuclearGhastMobKey(), PersistentDataType.BOOLEAN));

            if (!isNuclearGhast) {
                return;
            }

            YamlConfiguration configuration = (YamlConfiguration) plugin.getConfig();

            ConfigurationSection nuclearGhastDeathSettings = configuration.getConfigurationSection("nuclear-ghasts.death");
            assert nuclearGhastDeathSettings != null;

            ConfigurationSection explosionSettings = nuclearGhastDeathSettings.getConfigurationSection("poison");
            assert explosionSettings != null;

            boolean explosionEnabled = explosionSettings.getBoolean("explosion.enabled");

            World world = ghast.getWorld();
            Location location = ghast.getLocation();

            if (explosionEnabled) {
                world.createExplosion(location,
                        (float) explosionSettings.getDouble("power"),
                        explosionSettings.getBoolean("set-fire"),
                        explosionSettings.getBoolean("break-blocks"),
                        ghast
                );
            }

            ConfigurationSection poisonSettings = nuclearGhastDeathSettings.getConfigurationSection("poison");
            assert poisonSettings != null;

            boolean poisonEnabled = poisonSettings.getBoolean("enabled");

            if (poisonEnabled) {
                ThrownPotion potion = (ThrownPotion) world.spawnEntity(location, EntityType.POTION);

                potion.getEffects().add( // error
                        new PotionEffect(
                                PotionEffectType.POISON,
                                poisonSettings.getInt("duration-seconds") * 20,
                                poisonSettings.getInt("potency") - 1,
                                true,
                                true,
                                true
                        )
                );
            }
        }
    }
}

File: src/main/java/dev/enderman/minecraft/plugins/projectiles/better/listener/NuclearGhastDeathListener.kt
package dev.enderman.minecraft.plugins.projectiles.better.listener

import dev.enderman.minecraft.plugins.projectiles.better.BetterProjectilesPlugin
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.entity.EntityType
import org.bukkit.entity.Ghast
import org.bukkit.entity.ThrownPotion
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityDeathEvent
import org.bukkit.inventory.meta.PotionMeta
import org.bukkit.persistence.PersistentDataType
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType

class NuclearGhastDeathListener(private val plugin: BetterProjectilesPlugin) : Listener {

  @EventHandler
  fun onNuclearGhastDeath(event: EntityDeathEvent) {
    val entity = event.entity

    if (entity is Ghast) {
      val container = entity.persistentDataContainer

      val isNuclearGhast = true == container.get(plugin.nuclearGhastMobKey, PersistentDataType.BOOLEAN)

      if (!isNuclearGhast) return

      val configuration = plugin.config as YamlConfiguration

      val nuclearGhastDeathSettings = configuration.getConfigurationSection("nuclear-ghasts.death")!!

      val poisonSettings = nuclearGhastDeathSettings.getConfigurationSection("poison")!!
      val explosionSettings = nuclearGhastDeathSettings.getConfigurationSection("explosion")!!

      val explosionEnabled = explosionSettings.getBoolean("enabled")

      val world = entity.world
      val location = entity.location

      if (explosionEnabled) {
        plugin.logger.info("NUCLEAR GHAST EXPLOSION IMMINENT!")

        world.createExplosion(
          location,
          explosionSettings.getDouble("power").toFloat(),
          explosionSettings.getBoolean("set-fire"),
          explosionSettings.getBoolean("break-blocks"),
          entity
        )
      }

      val poisonEnabled = poisonSettings.getBoolean("enabled")

      if (poisonEnabled) {
        val potion = world.spawnEntity(location, EntityType.POTION) as ThrownPotion

        val item = potion.item
        val meta = item.itemMeta as PotionMeta

        meta.addCustomEffect(
          PotionEffect(
            PotionEffectType.POISON,
            poisonSettings.getInt("duration-seconds") * 20,
            poisonSettings.getInt("potency") - 1,
            true,
            true,
            true
          ),
          true
        )

        item.itemMeta = meta
        potion.item = item
      }
    }
  }
}
